security/s2a/internal/crypter/halfconn.go

```diff
@@ -0,0 +1,126 @@
+package crypter
+
+import (
+	"fmt"
+	"golang.org/x/crypto/cryptobyte"
+	s2a_proto "google.golang.org/grpc/security/s2a/internal"
+	"hash"
+	"sync"
+)
+
+const (
+	tls13Key    = "tls13 key"
+	tls13Nonce  = "tls13 iv"
+	tls13Update = "tls13 upd"
+)
+
+type S2AHalfConnection struct {
+	h             func() hash.Hash
+	aeadCrypter   s2aAeadCrypter
+	expander      hkdfExpander
+	seqCounter    counter
+	mutex         sync.Mutex
+	trafficSecret []byte
+	nonce         []byte
+}
+
+func NewHalfConn(ciphersuite s2a_proto.Ciphersuite, trafficSecret []byte) (S2AHalfConnection, error) {
+	cs := NewCiphersuite(ciphersuite)
+	if cs.trafficSecretSize() != len(trafficSecret) {
+		return S2AHalfConnection{}, fmt.Errorf("supplied traffic secret must be %v bytes, given: %v", cs.trafficSecretSize(), trafficSecret)
+	}
+
+	hc := S2AHalfConnection{h: cs.hashFunction(), expander: &defaultHKDFExpander{}, seqCounter: newCounter()}
+	key, err := hc.deriveSecret(hc.h, trafficSecret, []byte(tls13Key))
+	if err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("hc.deriveSecret(h, %v, %v) failed with error: %v", trafficSecret, tls13Key, err)
+	}
+	hc.aeadCrypter, err = cs.aeadCrypter(key)
+	if err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("cs.aeadCrypter(%v) failed with error: %v", key, err)
+	}
+	hc.nonce, err = hc.deriveSecret(hc.h, trafficSecret, []byte(tls13Nonce))
+	if err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("hc.deriveSecret(h, %v, %v) failed with error: %v", trafficSecret, tls13Nonce, err)
+	}
+	return hc, nil
+}
+
+func (hc *S2AHalfConnection) Encrypt(dst, plaintext, aad []byte) ([]byte, error) {
+	hc.mutex.Lock()
+	sequence, err := hc.getAndIncrementSequence()
+	if err != nil {
+		hc.mutex.Unlock()
+		return nil, err
+	}
+	nonce := hc.maskedNonce(sequence)
+	crypter := hc.aeadCrypter
+	hc.mutex.Unlock()
+	return crypter.encrypt(dst, plaintext, nonce, aad)
+}
+
+func (hc *S2AHalfConnection) Decrypt(dst, ciphertext, aad []byte) ([]byte, error) {
+	hc.mutex.Lock()
+	sequence, err := hc.getAndIncrementSequence()
+	if err != nil {
+		hc.mutex.Unlock()
+		return nil, err
+	}
+	nonce := hc.maskedNonce(sequence)
+	crypter := hc.aeadCrypter
+	hc.mutex.Unlock()
+	return crypter.decrypt(dst, ciphertext, nonce, aad)
+}
+
+func (hc *S2AHalfConnection) UpdateKey() error {
```

Comment:
The entire body of this function should be under the lock (to guard against concurrent calls to `Encrypt` and `UpdateKey`).